sdk/core/core-http/src/credentials/accessTokenRefresher.ts

```diff
@@ -2,29 +2,30 @@
 // Licensed under the MIT license.
 
 import { AccessToken, TokenCredential, GetTokenOptions } from "@azure/core-auth";
+import { delay } from "../util/utils";
+import { ExpiringAccessTokenCache } from "./accessTokenCache";
 
 /**
  * Helps the core-http token authentication policies with requesting a new token if we're not currently waiting for a new token.
  */
 export class AccessTokenRefresher {
   private promise: Promise<AccessToken | undefined> | undefined;
-  private lastCalled = 0;
+  private isTokenValid?: () => boolean;
 
   constructor(
     private credential: TokenCredential,
     private scopes: string | string[],
-    private requiredMillisecondsBeforeNewRefresh: number = 30000
-  ) {}
+    private tokenRefreshIntervalInMs: number = 30000
+  ) { }
 
   /**
-   * Returns true if the required milliseconds(defaulted to 30000) have been passed signifying
-   * that we are ready for a new refresh.
+   * Starts the refresher if we're in the refresh period.
    */
-  public isReady(): boolean {
-    // We're only ready for a new refresh if the required milliseconds have passed.
-    return (
-      !this.lastCalled || Date.now() - this.lastCalled > this.requiredMillisecondsBeforeNewRefresh
-    );
+  public startIfReady(options: GetTokenOptions, cache?: ExpiringAccessTokenCache): void {
+    if (!this.promise && cache?.isReadyToRefresh()) {
+      this.isTokenValid = cache.isTokenValid.bind(cache);
```

Comment:
this feels very strange to me, pulling a method off a class instance and saving a reference to it instead of to the instance itself. 

It also makes the state of the refresher extremely hard to reason about, since cache is optional and can replace `isTokenValid` at any time there isn't a promise already running.

Reply:
Resolving since I reverted this.